crates/saya/core/src/verifier/starknet.rs

```diff
@@ -2,73 +2,87 @@ use std::time::Duration;
 
 use anyhow::Context;
 use dojo_utils::{TransactionExt, TxnConfig};
+use itertools::Itertools;
 use starknet::accounts::{Account, Call, ConnectedAccount};
-use starknet::core::types::{Felt, TransactionExecutionStatus, TransactionStatus};
+use starknet::core::types::Felt;
 use starknet::core::utils::get_selector_from_name;
-use starknet::providers::Provider;
+use starknet_crypto::poseidon_hash_many;
 use tokio::time::sleep;
+use tracing::{info, trace};
 
-use crate::dojo_os::get_starknet_account;
-use crate::StarknetAccountData;
+use super::utils::wait_for_sent_transaction;
+use crate::{SayaStarknetAccount, LOG_TARGET};
+
+const CHUNK_SIZE: usize = 800;
 
 pub async fn starknet_verify(
     fact_registry_address: Felt,
     serialized_proof: Vec<Felt>,
-    starknet_config: StarknetAccountData,
+    cairo_version: Felt,
+    account: &SayaStarknetAccount,
 ) -> anyhow::Result<(String, Felt)> {
+    if serialized_proof.len() > CHUNK_SIZE {
+        trace!(target: LOG_TARGET,
+            "Calldata too long at: {} felts, transaction could fail, splitting it.",
+            serialized_proof.len()
+        );
+    }
+
     let txn_config = TxnConfig { wait: true, receipt: true, ..Default::default() };
-    let account = get_starknet_account(starknet_config)?;
-    let account = account.lock().await;
+
+    let mut nonce = account.get_nonce().await?;
+    let mut hashes = Vec::new();
+
+    for fragment in serialized_proof.into_iter().chunks(CHUNK_SIZE).into_iter() {
+        let mut fragment = fragment.collect::<Vec<_>>();
+        let hash = poseidon_hash_many(&fragment);
+        hashes.push(hash);
+
+        fragment.insert(0, fragment.len().into());
+
+        let tx = account
+            .execute_v1(vec![Call {
+                to: fact_registry_address,
+                selector: get_selector_from_name("publish_fragment").expect("invalid selector"),
+                calldata: fragment,
+            }])
+            .nonce(nonce)
+            // .max_fee(576834050002014927u64.into())
+            .send_with_cfg(&txn_config)
+            .await
+            .context("Failed to send `publish_fragment` transaction.")?;
+
+        wait_for_sent_transaction(tx.clone(), account).await?;
+
+        trace!(target: LOG_TARGET, "Sent `publish_fragment` transaction {:#x}", tx.transaction_hash);
+
+        sleep(Duration::from_secs(5)).await;
```

Comment:
We shouldn't have to wait 5 seconds if we're using a wait for transaction, that could add significant delay based on the number of fragments, could we check if this delay is actually required? Happy to debug with you if the wait is not effective.
But polling the status on a tx hash should result in no delay to executed subsequent tx.

Reply:
In the last commit I limited waits as much as I could without failing transactions. I now don't have to wait between write-only transactions.
I need to wait before the first - to fetch the correct nonce, and after the last. My theory at the moment is that the transaction is passing, but the state isn't updating until those 2 seconds later. 
I agree there should be no delay, but the transactions are failing nevertheless.

Reply:
Hum, weird, will have a look as we're supposed to poll the status. Of perhaps the account is not set on `pending` but `latest` instead.